src/main/java/it/pagopa/pm/gateway/controller/XPayPaymentController.java

```diff
@@ -255,58 +259,57 @@ private void executeXPayAuthorizationCall(AuthPaymentXPayRequest xPayRequest, Pa
         }
     }
 
-    private ResponseEntity<XPayPollingResponse> createXPayAuthPollingResponse(HttpStatus httpStatus, XPayPollingResponseError error, PaymentRequestEntity entity) {
+    private ResponseEntity<XPayPollingResponse> createXPayAuthPollingResponse(PaymentRequestEntity paymentRequestEntity) {
+        String requestId = paymentRequestEntity.getGuid();
+        PaymentRequestStatusEnum statusEnum = getEnumValueFromString(paymentRequestEntity.getStatus());
+        log.info("START - create XPay polling response for requestId {} - status {}", requestId, statusEnum);
+
         XPayPollingResponse response = new XPayPollingResponse();
+        response.setHtml(paymentRequestEntity.getXpayHtml());
+        response.setRequestId(requestId);
+        response.setPaymentRequestStatusEnum(statusEnum);
 
-        if (Objects.nonNull(error)) {
-            log.info("START - create XPay polling response - error case");
-            response.setError(error);
-            return ResponseEntity.status(httpStatus).body(response);
-        }
+        OutcomeXpayGateway outcomeXpayGateway = buildOutcomeXpayGateway(paymentRequestEntity.getErrorCode(),
+                paymentRequestEntity.getAuthorizationCode(), statusEnum);
+        response.setOutcomeXpayGateway(outcomeXpayGateway);
 
-        String requestId = entity.getGuid();
-        response.setRequestId(requestId);
+        if (isStatusOneOfGiven(statusEnum, AUTHORIZED, DENIED, CANCELLED)) {
+            ClientConfig clientConfig = clientsConfig.getByKey(paymentRequestEntity.getClientId());
+            String clientReturnUrl = clientConfig.getXpay().getClientReturnUrl();
+            response.setRedirectUrl(StringUtils.join(clientReturnUrl, paymentRequestEntity.getIdTransaction()));
+        }
 
-        log.info("START - create XPay polling response for requestId: " + requestId);
-        String status = entity.getStatus();
-        log.info(String.format("Request status for requestId %s is %s", requestId, status));
+        log.info("END - create XPay polling response for requestId " + requestId);
+        return ResponseEntity.ok().body(response);
+    }
 
-        String clientReturnUrl = clientsConfig.getByKey(entity.getClientId()).getXpay().getClientReturnUrl();
+    private boolean isStatusOneOfGiven(PaymentRequestStatusEnum referenceStatus, PaymentRequestStatusEnum... given) {
+        List<PaymentRequestStatusEnum> paymentRequestStatusEnums = Arrays.asList(given);
+        return paymentRequestStatusEnums.contains(referenceStatus);
+    }
```

Comment:
As declared this function can accept even zero or one `PaymentRequestStatusEnum`s. For these such a method doesn't make sense. I'd suggest changing the method to:

```suggestion
    private boolean isStatusOneOfGiven(PaymentRequestStatusEnum referenceStatus, PaymentRequestStatusEnum option1, PaymentRequestStatusEnum option2, PaymentRequestStatusEnum... otherOptions) {
        Set<PaymentRequestStatusEnum> paymentRequestStatusEnums = Arrays.stream(otherOptions).collect(Collectors.toSet());
        paymentRequestStatusEnums.add(option1);
        paymentRequestStatusEnums.add(option2);

        return paymentRequestStatusEnums.contains(referenceStatus);
    }
```

Reply:
Also (bikeshedding): `isStatusOneOfGiven` -> `isStatusOneOf`